### PR TITLE
Use symfony translation service

### PR DESCRIPTION
**Pull request by gimler**
I have rewrite the code to use the symfony translation service. So every translation component format is now supported for im and export. Every file is exported back to his original file and format.

Also nested yml syntax is supported now.

Now it is much more cleaner i reduce the code to 10% ;)

### CODE REVIEW

File: Command/ExportCommand.php [hunk 1]
```diff
@@ -155,21 +156,21 @@ protected function exportLocale($bundle, $locale, $entryList)
      */
     protected function exportTranslationFileList($bundle, $locale, ArrayCollection $translationFileList)
     {
-        $exporterService = $this->getContainer()->get('server_grove_translation_editor.exporter');
-
-        $fileExtension = $exporterService->getFileExtension();
-        $localeString  = (string) $locale;
+        $writer    = $this->getContainer()->get('translation.writer');
```
Comment (guilhermeblanco) [inline Command/ExportCommand.php]: Extra spaces between variable name and `=`.

File: Command/ExportCommand.php [hunk 2]
```diff
@@ -155,21 +156,21 @@ protected function exportLocale($bundle, $locale, $entryList)
...
 
         foreach ($translationFileList as $fileName => $translationList) {
-            // Generating file name
-            $fileName = sprintf('%s.%s.%s', $fileName, $localeString, $fileExtension);
             $filePath = sprintf('%s/%s/%s', $bundle->getPath(), self::TRANSLATION_PATH, $fileName);
 
             $this->output->write(sprintf('    Exporting "<info>%s</info>"... ', $fileName));
 
-            $exportResult = $exporterService->exportFile($filePath, $translationList);
+            $catalogue = new MessageCatalogue((string) $locale);
```
Comment (guilhermeblanco) [inline Command/ExportCommand.php:166]: Missing line break

File: Command/ImportCommand.php [hunk 3]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+            $storage = $this->getContainer()->get('server_grove_translation_editor.storage');
+            $loader  = $this->getContainer()->get('translation.loader');
 
-            $locale = $importerService->importLocale($language, $country);
-            $importerService->importFile($bundle, $locale, $translationFilePath);
+            // get or create locale
+            $localeList = $storage->findLocaleList(array(
+                'language' => $properties['language'],
+                'country'  => $properties['country']
+            ));
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php:108]: Missing line break

File: Command/ImportCommand.php [hunk 4]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+            if (null === $locale = array_shift($localeList)) {
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php:110]: Do not use yoda conditions.

``` php
if (($locale = array_shift($localeList)) === null) {
```

File: Command/ImportCommand.php [hunk 5]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+                $locale = $storage->createLocale($properties['language'], $properties['country']);
+            }
+
+            // import translation
+            $entryList = $storage->findEntryList(array(
+                'domain'   => $bundle->getName(),
+                'fileName' => $properties['name']
+            ));
+
+            $catalogue = new MessageCatalogue($locale);
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php:120]: Missing line break

File: Command/ImportCommand.php [hunk 6]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+            $loader->loadMessages(dirname($translationFilePath), $catalogue);
+
+            foreach ($catalogue->all($properties['name']) as $alias => $value) {
+                // get or create entry
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php]: Always start a comment with upper cased chars. =)

File: Command/ImportCommand.php [hunk 7]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+                    $entryList,
+                    function ($entry) use ($alias) {
+                        return ($alias === $entry->getAlias());
+                    }
+                );
+                if (null === $entry = array_shift($entryListFiltered)) {
+                    $entry = $storage->createEntry($bundle->getName(), $properties['name'], pathinfo($translationFilePath, PATHINFO_EXTENSION),$alias);
+                }
+
+                // get or create translation
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php]: Always start a comment with upper cased chars. =)

File: Command/ImportCommand.php [hunk 8]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+
+            foreach ($catalogue->all($properties['name']) as $alias => $value) {
+                // get or create entry
+                $entryListFiltered = array_filter(
+                    $entryList,
+                    function ($entry) use ($alias) {
+                        return ($alias === $entry->getAlias());
+                    }
+                );
+                if (null === $entry = array_shift($entryListFiltered)) {
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php:133]: Missing line break.
Avoid yoda conditions.
Encapsulate attribution and comparison

File: Command/ImportCommand.php [hunk 9]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+                    $entry = $storage->createEntry($bundle->getName(), $properties['name'], pathinfo($translationFilePath, PATHINFO_EXTENSION),$alias);
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php]: Missing space after comma.
Also, break this call into multiple lines.

File: Command/ImportCommand.php [hunk 10]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+                }
+
+                // get or create translation
+                $translationList = $entry->getTranslations()->filter(
+                    function ($translation) use ($locale) {
+                        return ($translation->getLocale() === $locale);
+                    }
+                );
+                if (0 === count($translationList)) {
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php:149]: Missing line break

File: Command/ImportCommand.php [hunk 11]
```diff
@@ -129,7 +167,7 @@ protected function getTranslationFileList($bundle, $filterLocaleName = null, $fi
         }
 
         $finder = new Finder();
-        $finder->files()->in($translationPath)->name('*');
+        $finder->files()->in($translationPath)->notName('*~');
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php]: Missing inline comment here.

_Personal comment:_ WTH is "*~"

File: Command/ImportCommand.php [hunk 12]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
...
+            // get or create locale
+            $localeList = $storage->findLocaleList(array(
+                'language' => $properties['language'],
+                'country'  => $properties['country']
+            ));
+            if (null === $locale = array_shift($localeList)) {
+                $locale = $storage->createLocale($properties['language'], $properties['country']);
+            }
+
+            // import translation
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php]: Always start a comment with upper cased chars. =)

File: Command/ImportCommand.php [hunk 13]
```diff
@@ -94,15 +95,52 @@ protected function importBundle($bundle)
         }
 
         // Importing files
-        $importerService = $this->getContainer()->get('server_grove_translation_editor.importer');
-
-        foreach ($translationFileList as $translationFilePath) {
+        foreach ($translationFileList as $translationFilePath => $properties) {
```
Comment (guilhermeblanco) [inline Command/ImportCommand.php]: Never user plural words ($properties). They confuse developers.
Prefer a suffix line *List.

File: Command/ImportCommand.php [hunk 14]
```diff
@@ -146,7 +184,12 @@ protected function getTranslationFileList($bundle, $filterLocaleName = null, $fi
                 continue;
             }
 
-            $translationFiles[] = $translationFilePath;
+            $translationFiles[$translationFilePath] = array(
```
Comment (robocoder) [inline Command/ImportCommand.php:196]: The original didn't use $translationFileList ?

File: Storage/AbstractStorage.php [hunk 15]
```diff
@@ -44,13 +44,14 @@ public function createLocale($language, $country = null)
     /**
      * {@inheritdoc}
```
Comment (robocoder) [inline Storage/AbstractStorage.php:45]: The original had the wrong phpdoc for an abstract base class that doesn't extend/implement anything...

**Comment by guilhermeblanco** [comment]
@gimler can you work on these changes I commented?
I'd be happy to merge your PR once they're done. ATM our team is focused on a different part of the application.

**Comment by gimler** [comment]
I will fix this issues today

File: Command/ExportCommand.php (hunk 1 above)
Comment (gimler) [inline Command/ExportCommand.php]: fixed

File: Command/ExportCommand.php (hunk 2 above)
Comment (gimler) [inline Command/ExportCommand.php:166]: fixed

File: Command/ImportCommand.php (hunk 3 above)
Comment (gimler) [inline Command/ImportCommand.php:108]: fixed

File: Command/ImportCommand.php (hunk 5 above)
Comment (gimler) [inline Command/ImportCommand.php:120]: fixed

File: Command/ImportCommand.php (hunk 9 above)
Comment (gimler) [inline Command/ImportCommand.php]: fixed

File: Command/ImportCommand.php (hunk 10 above)
Comment (gimler) [inline Command/ImportCommand.php:149]: fixed

File: Command/ImportCommand.php (hunk 11 above)
Comment (gimler) [inline Command/ImportCommand.php]: fixed

File: Command/ImportCommand.php (hunk 13 above)
Comment (gimler) [inline Command/ImportCommand.php]: fixed

File: Command/ImportCommand.php (hunk 6 above)
Comment (gimler) [inline Command/ImportCommand.php]: fixed

File: Command/ImportCommand.php (hunk 7 above)
Comment (gimler) [inline Command/ImportCommand.php]: fixed

File: Command/ImportCommand.php (hunk 12 above)
Comment (gimler) [inline Command/ImportCommand.php]: fixed

File: Command/ImportCommand.php (hunk 4 above)
Comment (gimler) [inline Command/ImportCommand.php:110]: why the symfony coding standard use yoda conditions also it protect you.

I can move the array_shift out of the condition?

File: Command/ImportCommand.php (hunk 8 above)
Comment (gimler) [inline Command/ImportCommand.php:133]: > Missing line break.
> fixed
> 
> Avoid yoda conditions.
> see other code comment

File: Command/ImportCommand.php (hunk 14 above)
Comment (gimler) [inline Command/ImportCommand.php:196]: What do you mean? I add the file path as key so i need it later. And so i reduce the code.

File: Storage/AbstractStorage.php (hunk 15 above)
Comment (gimler) [inline Storage/AbstractStorage.php:45]: right but this is not part of this PR so i would do this for all functions in a seperate PR

File: Command/ImportCommand.php (hunk 4 above)
Comment (guilhermeblanco) [inline Command/ImportCommand.php:110]: Whenever you can leave the assignments out of the conditional expression, it's better.

File: Command/ImportCommand.php (hunk 14 above)
Comment (guilhermeblanco) [inline Command/ImportCommand.php:196]: He was wondering if the original variable name was `$translationFiles`or `$translationFileList`.
If you can change to `$translationFileList`, it's gonna be better.

**Comment by guilhermeblanco** [comment]
@gimler thanks a lot for your efforts on this one.
I accepted the PR because it's already in a very good shape.
My suggestions can be applied on another PR. =)
